### PR TITLE
RFC: sets/domains

### PR DESCRIPTION
**Pull request by tbreloff**
What do you think? Will this work for our purposes?

### CODE REVIEW

**Comment by tbreloff** [comment]
We might consider having an `outputdomain` in addition, for example with SVM isn't the output domain an `IntervalSet(-1,1)` but the target domain is `DiscreteSet([1,-1])`?  Though I guess the more correct way to do this is that the margin loss has `inputdomain(loss) == TupleSet(IntervalSet(-1,1), DiscreteSet([1,-1]))`.

Anyways... criticism welcome

**Comment by Evizero** [comment]
> for example with SVM isn't the output domain an IntervalSet(-1,1)

No. The model is just a linear model like in linear regression. the interpretation is just different. So the output domain is R

EDIT:
The points that are not filled in in this plot will have an output greater than 1 or smaller than -1. all points between the dotted lines have an output in [-1,1]

svm

File: src/LearnBase.jl [hunk 1]
```diff
@@ -117,6 +117,60 @@ function update! end
...
+    hi::T
+end
+
+function IntervalSet{A<:Number,B<:Number}(lo::A, hi::B)
+    T = promote_type(A,B)
+    IntervalSet{T}(convert(T,lo), convert(T,hi))
+end
+
+Base.rand{T<:Number}(s::IntervalSet{T}) = rand() * (s.hi - s.lo) + s.lo
+Base.in{T<:Number}(x::Number, s::IntervalSet{T}) = s.lo <= x <= s.hi
```
Comment (Evizero) [inline src/LearnBase.jl:138]: can you please add some tests for `in` and `rand`

**Comment by Evizero** [comment]
I'll be greedy and ask for some tests. Other than that I like it 👍

**Comment by tbreloff** [comment]
Yeah was planning on tests.  I just wanted opinions before I went to the trouble.

**Comment by tbreloff** [comment]
Cleaned up and tests added.  I'd be happy to merge if you are.  cc: @Evizero @ahwillia @spencerlyon2

**Comment by Evizero** [comment]
Looks good to me in terms of content.

But before we merge, please change the tabs to 4 spaces in the tests :)

File: src/LearnBase.jl [hunk 2]
```diff
@@ -117,6 +117,72 @@ function update! end
...
+
+
+"Groups several heterogenous sets.  Used mainly for proper dispatch."
+immutable TupleSet{T<:Tuple} <: AbstractSet
+    sets::T
+end
+TupleSet(sets::AbstractSet...) = TupleSet(sets)
+Base.rand(sets::TupleSet) = [rand(s) for s in sets.sets]
+function Base.rand(sets::TupleSet, dim1::Integer, dims::Integer...)
+    A = Array(typeof(rand(sets)), dim1, dims...)
```
Comment (sglyon) [inline src/LearnBase.jl]: Probably super minor, but you could save the first `rand(sets)` if you wanted

**Comment by sglyon** [comment]
Sorry I didn't pitch in earlier -- I didn't have much to add. 

I would vote that we add a method for generating one sample from a `TupleSet` and put the output in a `Tuple` similar to what we did [here](https://github.com/tbreloff/Reinforce.jl/blob/da0dbfccfe8a6f4ee5427d628a1f67047a8884c4/src/Reinforce.jl#L84)

Otherwise, looks good to me!

**Comment by tbreloff** [comment]
> you could save the first rand(sets) if you wanted

 Thought about this, though it makes the code uglier.  I think a better solution is to have a query method to static-ly return the type that rand will output.  Something like:

``` julia
randtype(s::IntervalSet) = Float64
```

and then use this to compute the randtype from the randtype of its components.

> I would vote that we add a method for generating one sample from a TupleSet and put the output in a Tuple similar to what we did here

Ok

Also... tests are failing on 0.4 because of types returned from array comprehensions.  What do we want to do?

```
ERROR: LoadError: test failed: Array{Union{Float64,Int64},1} == Array{Real,1}
```

**Comment by tbreloff** [comment]
> What do we want to do?

One option is that we don't support 0.4.  Just one more motivating factor to get everyone on 0.5 as soon as possible.

**Comment by Evizero** [comment]
if travis succeeds go ahead and merge. I just enabled appveyor, so if that one fails it may be for other reasons than this PR

**Comment by Evizero** [comment]
looks like appveyor just succeeded on master, so it should hopefully succeed here as well

**Comment by sglyon** [comment]
Should `randtype` be `Base.eltype`?

**Comment by tbreloff** [comment]
Needs more review.  I added a method `randtype` which, as it sounds, returns the type that will be returned by `rand`.  I also went ahead and changed the logic to use group promotion so that we will create concrete types even when combining different set types.  For example:

``` julia
s = [IntervalSet(0,1), DiscreteSet([0,1])]
@test typeof(rand(s)) == Vector{Float64}
@test typeof(rand(s, 2)) == Vector{Vector{Float64}}
```

> Should randtype be eltype?

No.  See for example `IntervalSet(0,1)`

**Comment by Evizero** [comment]
lgtm

**Comment by Evizero** [comment]
nice thanks. Since this is now more than just definitions I should probably add coveralls as well

**Pull request merged**